spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TempDirInterceptor.java

```diff
@@ -0,0 +1,112 @@
+package org.spockframework.runtime.extension.builtin;
+
+import org.codehaus.groovy.runtime.ResourceGroovyMethods;
+import org.spockframework.runtime.extension.IMethodInterceptor;
+import org.spockframework.runtime.extension.IMethodInvocation;
+import org.spockframework.runtime.model.FieldInfo;
+import org.spockframework.runtime.model.MethodKind;
+import org.spockframework.util.Beta;
+
+import java.io.IOException;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.attribute.BasicFileAttributes;
+
+import static java.nio.file.FileVisitResult.CONTINUE;
+
+/**
+ * @author dqyuan
+ * @since 2.0
+ */
+@Beta
+public class TempDirInterceptor implements IMethodInterceptor {
+  private static final String TEMP_DIR_PREFIX = "spock";
+
+  private final Class<?> fieldType;
+  private final FieldInfo fieldInfo;
+  private final Path parentDir;
+  private final MethodKind interceptPoint;
+  private final boolean keep;
+
+  private Path tempPath;
+
+  TempDirInterceptor(Class<?> fieldType, FieldInfo fieldInfo,
+                     MethodKind interceptPoint, Path parentDir,
+                     boolean keep) {
+    this.fieldType = fieldType;
+    this.fieldInfo = fieldInfo;
+    this.parentDir = parentDir;
+    this.interceptPoint = interceptPoint;
+    this.keep = keep;
+  }
+
+  private Path generateTempDir(IMethodInvocation invocation) throws IOException {
+    String prefix = TEMP_DIR_PREFIX + "_" + ((invocation.getIteration() == null) ?
```

Comment:
This can lead to some long names, why dont we use a uuid or currentTimeMillis ?

Reply:
That was my suggestion.
This way you can easily identify to which test a temp directory belongs, especially if you use the `keep` option.

Reply:
If you don't like it, we can of course switch back to `Files.createTempDirectory(TEMP_DIR_PREFIX);`

Reply:
I think it is only a prefix, no need to be long.

Reply:
The intention by me was not for it to be long, as the method will generate unique names either way.
The intention by me was to be able to see which temp dir belongs to which test.

Reply:
I would rather check for the length, and shorten if it is longer than, e.g. 30chars, if you want you could make that configurable. If it is an iteration you could use substring of the name + iteration index.

Windows has a limit of 260 chars for a full path, unless you enabled long path names.

Reply:
Hm, yeah, Windows path length restrictions are a very good point.
But imho this is not worth a configuration option.
The directories will anyway still be unique, as the calculated string is only used as prefix.
So how about this version:
```java
private Path generateTempDir(IMethodInvocation invocation) throws IOException {
  StringBuilder prefix = new StringBuilder(TEMP_DIR_PREFIX);
  prefix.append('_');
  prefix.append(((invocation.getIteration() == null)
    ? invocation.getSpec().getName()
    : invocation.getIteration().getName())
    .replaceAll("[^a-zA-Z0-9_.-]++", "_"));
  if (prefix.length() > 25) {
    prefix.setLength(25);
  }
  if (invocation.getIteration() != null) {
    prefix.append('_').append(invocation.getIteration().getIterationIndex());
  }
  prefix.append('_').append(fieldInfo.getName());

  if (parentDir == null) {
    return Files.createTempDirectory(prefix.toString());
  }
  return Files.createTempDirectory(parentDir, prefix.toString());
}
```

Reply:
looks good